ボウリングのスコア計算プログラムオブジェクト指向版

naokinaokiboo opened the pull request:
「ボウリングのスコア計算プログラムオブジェクト指向版」の課題内容をpushしました。
レビューよろしくお願い致します。🙇‍♂️

naokinaokiboo committed:
FrameクラスとTestFrameクラスのひな形を追加

naokinaokiboo committed:
FrameクラスとTestFrameクラスの実装を追加

naokinaokiboo committed:
Gameクラス、TestGameクラスのひな形を追加

naokinaokiboo committed:
Gameクラス、TestGameクラスにFrameオブジェクトを生成する実装を追加

naokinaokiboo committed:
Frameクラスの不要なアクセサーを削除

それに伴い
・テストコードを修正
・GameクラスでのFrameオブジェクト生成時のコードを修正

naokinaokiboo committed:
Gameクラス、TestGameクラスにスコア計算の実装を追加

naokinaokiboo committed:
コマンドラインから引数を指定できるように変更(bowling.rbを追加)

naokinaokiboo committed:
メソッド名を変更(first_pins -> first_shot, second_pins -> second_shot)

naokinaokiboo committed:
rubocopの指摘に対応

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

naokinaokiboo committed:
不要なテストを削除

naokinaokiboo committed:
test_frame.rbのアサーションを変更（リファクタリング）

・nilの確認：assert_nilを使用
・真偽値の確認：assert, refuteを使用

naokinaokiboo committed:
定数をクラス内に移動

naokinaokiboo committed:
eachメソッドでの配列への要素追加をmapメソッドに置き換え（リファクタリング）

naokinaokiboo committed:
メソッド「join_to_last_frame」内部の処理を定数利用から固定値に変更（リファクタリング）

・ボーリングのゲーム数に関するものであり、固定値のほうが可読性が高いと判断したため

naokinaokiboo committed:
テストコード内の不要なsetupメソッドを削除

naokinaokiboo committed:
フレーム毎にスコアマークを配列にまとめる処理で、不要なメソッド化と場合分けを削除

・メソッド化をやめた理由
　メソッド名で処理の意図を伝えるためにメソッド化していたが、そこまでするほどの処理では無かったため
・場合分けをやめた理由
　場合分けせずとも、同一の処理で同じ結果になるため
　（スコアの条件次第では、不要な処理になる場合もあるが、この方が可読性が高まると判断）

naokinaokiboo committed:
スコアマークがストライクかどうかを判断するメソッドを追加

naokinaokiboo committed:
injectメソッドをsumメソッドで置き換え

naokinaokiboo committed:
メソッド「calculate_frame_score」の仮引数の書き方を変更

・丸括弧の使用を中止（メソッド使用者からするとトリッキーすぎるため）

naokinaokiboo committed:
Gameクラスのinitializeで行っていたFrameクラスの生成をcalculate_total_scoreに移動

naokinaokiboo committed:
長いメソッド名・変数名を短く修正

naokinaokiboo committed:
不要な空行を削除

naokinaokiboo committed:
Gameクラスのinitializeの処理をcalculate_total_scoreへ移動（以前のコミットでの移動漏れ）

naokinaokiboo commented:
レビューありがとうございます！
動画でご指摘・ご説明頂いた点について、明らかに直したい点について修正してみました。
再度ご確認よろしくお願い致します。🙇‍♂️

また、動画の中で`スコア計算の部分をもう少しオブジェクト指向っぽく・・・`ということも言及されていたので、自分なりにどう考えていたかをコメントしてみます。

- 当初、ストライク、スペア時のボーナスも含めて、各フレームに自フレームのスコアを計算させるという方針も案としてはありました。（その場合、対象のフレームに次のフレーム、さらにその次のフレームを渡すしかないかな・・・と思っています）
- 一方で、今回は不要ですがボーリングのスコア計算には新ルールもあることから、スコア計算の部分を切り出すという案も思い浮かびました。（各フレームをもらってスコア計算を担うクラス（またはモジュール））

ただ、
- 本プラクティスでは新ルールは不要
- 不要なのであれば、設計・実装は先延ばしにするべき（先延ばしというかこのプラクティスで実装することはないのですが・・・）

という考えも出てきて切り分けるのもどうかと思い、Gameクラスが各フレームを管理しスコア計算するというカタチになりました。😅

なかなか自分の中でも、「〜な実装がベストだ！」とはっきり言い切れる考えがなく、とりとめの無い文章になってしまいましたが、もし何かあればコメントいただけると有り難いです。🙇‍♂️

JunichiIto commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,60 @@
+
+  def generate_marks_by_frame(marks)
+    marks_by_strike_or_not =
+      marks.slice_when { |left, right| Frame.strike_mark?(left) || Frame.strike_mark?(right) }
```
rightの条件はなくしてもうまく動きそうです。

naokinaokiboo replied:
この後で`each_slice(2)`をすることを考えると、確かに"X"の後ろで区切っておけば、問題無かったですね。。。😅
rightの条件は削除致します。

JunichiIto commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,60 @@
+require_relative './frame'
+
+class Game
+  NUM_OF_FRAMES = 10
```
人間の目に触れる箇所は1〜10とすれば良いですが、プログラム内では0〜9で扱った方がいろいろと都合が良さそうです。
今回はフレーム番号を人間に見せるケースはないため、常に0〜9でいいんじゃないかなと思います。

もしフレーム番号を見せる場合があったとしても、「プログラム内では原則0〜9で扱い、表示するときに+1する」という実装にする方がコードがシンプルになると思います。

あと、（若干微妙なところですが）「ボウリングは10フレームであることは半ば常識（1週間は7日であることをみんな知ってるのと同じ）」とすれば、わざわざ定数化しなくてもいいんじゃないかなーと思います。

naokinaokiboo replied:
> あと、（若干微妙なところですが）「ボウリングは10フレームであることは半ば常識（1週間は7日であることをみんな知ってるのと同じ）」とすれば、わざわざ定数化しなくてもいいんじゃないかなーと思います。

既にフレームに関する数字がプログラム中に出てきていることもあるので、今回は定数化しない方向で修正します。

JunichiIto commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,60 @@
+      Frame.new(marks_in_frame)
+    end
+
+    (0..NUM_OF_FRAMES - 1).sum do |frame_idx|
+      calculate_frame_score(frame_idx + 1, *frames[frame_idx, 3])
+    end
```
```suggestion
    frames.each_with_index.sum do |frame, i|
      next_frame = ...
      after_next_frame = ...
      frame.calculate_score(next_frame, after_next_frame)
    end
```

みたいにしてFrameクラスにスコア計算させた方がオブジェクト指向らしくなりそうです。

JunichiIto commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,37 @@
+    @shots[0] + @shots[1] == MAX_PINS
+  end
+
+  def self.strike_mark?(mark)
```
クラスメソッドはinitializeの前に定義することが多いかもしれません。
（特に決まりがあるわけでは無いですが）

JunichiIto reviewed (commented):
コメントしました〜。

naokinaokiboo committed:
slice_whenの不要な条件を削除

naokinaokiboo committed:
ボウリングは10フレームであることを常識として、定数化をやめる

naokinaokiboo committed:
フレームNo.の使用をやめて、フレームindexに統一

naokinaokiboo committed:
各フレームのスコア計算をFrameクラスで行うように変更

naokinaokiboo committed:
クラスメソッドをinitializeの前に移動

naokinaokiboo committed:
変数名を変更（リファクタリング）

JunichiIto commented on 06.bowling_object/game.rb:
```diff
@@ -15,46 +13,22 @@ def calculate_total_score
-    marks_by_strike_or_not =
-      marks.slice_when { |left, right| Frame.strike_mark?(left) || Frame.strike_mark?(right) }
+    marks_sliced_after_strike =
+      marks.slice_when { |left, _right| Frame.strike_mark?(left) }
```
slice_when以外のメソッドを使うと`_right`を無くせるかもしれません。

JunichiIto commented on 06.bowling_object/frame.rb:
```diff
@@ -3,6 +3,10 @@
 class Frame
```
計算処理をGameからFrameに持ってきたので、いくつかの処理をpublicメソッドにしなくても良くなってます。

naokinaokiboo replied:
こちら、確認不足でした。💦
適切なアクセス制限を設定するように致します。

JunichiIto reviewed (commented):
コメントしました〜。

naokinaokiboo committed:
slice_when -> slice_afterに変更（不要なブロック変数を削除できるため）

naokinaokiboo committed:
Frameクラスのメソッドに適切なアクセス制限を設定

naokinaokiboo committed:
アクセス制限の設定によって通らなくなったテストコードをメンテナンス

JunichiIto commented on 06.bowling_object/frame.rb:
```diff
@@ -13,6 +13,20 @@ def initialize(score_marks)
+    end
+  end
+
+  protected
```
ちゃんとprotectedの存在に気づけてて偉い 👍

JunichiIto commented on 06.bowling_object/test/test_frame.rb:
```diff
@@ -6,102 +6,102 @@
   test '#base_scoreはスペアでもストライクでも無い場合、1投目と2投目の合計を返す' do
     frame1 = Frame.new(%w[0 0])
-    assert_equal 0, frame1.base_score
+    assert_equal 0, frame1.send(:base_score)
```
https://github.com/naokinaokiboo/ruby-practices/pull/10/commits/4fd8f1c0a5311d6f89a5009c403240410e18229b と同じ話になりますが、publicじゃないメソッドをテストする必要はあるのか？という議論になりそうです。

publicなメソッドA が private (or protected)なB を呼びだしている場合は、「メソッドAを経由してメソッドBをテストする」のが原則です。
「どうしても」という場合はメソッドBをsendメソッド等でテストすることはまれにありますが、今回がその「どうしても」に該当するのかどうか、検討が必要かなーと思います。

naokinaokiboo replied:
そうですね。今回も`test/test_game.rb`のテストが通っていれば問題ないので、削除しておきます。😅

あと、Frameクラスのpublicメソッドのテストについて１点確認させてください。🙏
- publicメソッドについては、原則テストすべき
とは思いつつも
- 今回の場合、`test/test_game.rb`でテストしているルート以外では使用されません。

よって、テストコードを追加すると冗長になるかと考え追加していないのですが、どうなのでしょうか。💦

JunichiIto replied:
よい質問ですね！

「純粋な単体テスト」では対象のクラスのみをテスト対象として、それ以外のクラスはモックにします。
つまり、Gameクラスをテストする場合はFrameクラスをモックにしてFrameクラス本体は呼び出さないようにするわけです。
この場合、FrameクラスはFrameクラスで単体テストを行う必要が出てきます。

が、現実問題としてそこまでやるのは非常に面倒です😅
今回のように、Gameクラスをテストしたら自動的にFrameクラスもテストしたことになる、というケースは結構あります。
絶対的な正解があるわけではありませんが、こういう場合、僕はGameクラスのテストだけで十分としています。
もしかすると人によってはそれぞれテストすべき、という人がいるかもしれません。（あまりいないかな〜？）

というわけで今回はGameクラスのテストだけで良いと思います〜。

JunichiIto reviewed (commented):
コメントしました〜。

naokinaokiboo committed:
private,protectedなメソッドの冗長なテストを削除

JunichiIto commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+class Frame
+  MAX_PINS = 10
```
sで終わると配列かな？と勘違いされる可能性があります。
そもそもの話として、この値もはたして定数化する必要があるのか？と考えてみるといいかもしれません。

JunichiIto commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,67 @@
+  private
+
+  def spare?
+    return false if strike?
+
+    @shots[0] + @shots[1] == MAX_PINS
```
```suggestion
    !strike? && @shots[0..1].sum == MAX_PINS
```

みたいな書き方もできます。

JunichiIto reviewed (approved):
数点コメントしましたが、大きな問題ではないのでこれで合格とします〜 🙆‍♂️

naokinaokiboo committed:
ボーリングのピンが10本であることは自明とし、定数化をやめる

naokinaokiboo committed:
spare?の戻り値を1行にまとめる
